libs/vkd3d/command.c

```diff
@@ -4929,16 +4929,65 @@ static bool vk_image_memory_barrier_from_d3d12_transition(const struct d3d12_dev
     return vk_barrier->oldLayout != vk_barrier->newLayout;
 }
 
+static void vk_image_memory_barrier_for_after_aliasing_barrier(struct d3d12_device *device,
+        struct d3d12_resource *after, VkImageMemoryBarrier *vk_barrier)
+{
+    const struct vkd3d_format *vk_format = vkd3d_get_format(device, after->desc.Format, false);
+
+    vk_barrier->sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    vk_barrier->pNext = NULL;
+    vk_barrier->srcAccessMask = 0;
+    vk_barrier->dstAccessMask = VK_ACCESS_MEMORY_WRITE_BIT | VK_ACCESS_MEMORY_READ_BIT;
+    vk_barrier->oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+    vk_barrier->newLayout = after->common_layout;
+    vk_barrier->srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    vk_barrier->dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    vk_barrier->image = after->vk_image;
+    vk_barrier->subresourceRange.aspectMask = vk_format->vk_aspect_mask;
+    vk_barrier->subresourceRange.baseMipLevel = 0;
+    vk_barrier->subresourceRange.levelCount = VK_REMAINING_MIP_LEVELS;
+    vk_barrier->subresourceRange.baseArrayLayer = 0;
+    vk_barrier->subresourceRange.layerCount = VK_REMAINING_ARRAY_LAYERS;
+}
+
+static bool d3d12_resource_may_alias_other_resources(struct d3d12_resource *resource)
+{
+    /* Treat a NULL resource as "all" resources. */
+    if (!resource)
+        return true;
+
+    /* Cannot alias if the resource is allocated in a dedicated heap. */
+    if (resource->flags & VKD3D_RESOURCE_DEDICATED_HEAP)
+        return false;
+
+    /* Cannot alias if the resource is a swapchain image.
+     * This is also important since we cannot reason about a default
+     * image layout for presentable images like we can other resources. */
+    if (d3d12_resource_is_texture(resource) && (resource->flags & VKD3D_RESOURCE_PRESENT_STATE_TRANSITION))
+        return false;
+
+    return true;
+}
+
 static void STDMETHODCALLTYPE d3d12_command_list_ResourceBarrier(d3d12_command_list_iface *iface,
         UINT barrier_count, const D3D12_RESOURCE_BARRIER *barriers)
 {
     struct d3d12_command_list *list = impl_from_ID3D12GraphicsCommandList(iface);
     const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
-    bool have_aliasing_barriers = false, have_split_barriers = false;
+    bool have_split_barriers = false;
     VkPipelineStageFlags dst_stage_mask, src_stage_mask;
+    VkImageMemoryBarrier vk_image_barriers_stack[64];
+    VkImageMemoryBarrier *vk_image_barriers;
     VkImageMemoryBarrier vk_image_barrier;
     VkMemoryBarrier vk_memory_barrier;
-    unsigned int i;
+    unsigned int num_image_barriers;
+    unsigned int i, j;
+
+    num_image_barriers = 0;
+    if (barrier_count > ARRAY_SIZE(vk_image_barriers_stack))
+        vk_image_barriers = vkd3d_malloc(barrier_count * sizeof(*vk_image_barriers));
```

Comment:
Not a big fan of the allocation here; can't we just emit image barriers one by one? They should be very rare in practice.

Reply:
Fixed.